tests/test_petsc.py

```diff
@@ -0,0 +1,57 @@
+from devito import Grid
+from devito.ir.iet import Call, ElementalFunction, Definition, DummyExpr
+from devito.passes.iet.languages.C import CDataManager
+from devito.types import (DM, Mat, Vec, PetscMPIInt, KSP,
+                          PC, KSPConvergedReason, PETScFunction)
+import numpy as np
+
+
+def test_petsc_local_object():
+    """
+    Test C++ support for PETSc LocalObjects.
+    """
+    lo0 = DM('da')
+    lo1 = Mat('A')
+    lo2 = Vec('x')
+    lo3 = PetscMPIInt('size')
+    lo4 = KSP('ksp')
+    lo5 = PC('pc')
+    lo6 = KSPConvergedReason('reason')
+
+    iet = Call('foo', [lo0, lo1, lo2, lo3, lo4, lo5, lo6])
+    iet = ElementalFunction('foo', iet, parameters=())
+
+    dm = CDataManager(sregistry=None)
+    iet = CDataManager.place_definitions.__wrapped__(dm, iet)[0]
+
+    assert 'DM da;' in str(iet)
+    assert 'Mat A;' in str(iet)
+    assert 'Vec x;' in str(iet)
+    assert 'PetscMPIInt size;' in str(iet)
+    assert 'KSP ksp;' in str(iet)
+    assert 'PC pc;' in str(iet)
+    assert 'KSPConvergedReason reason;' in str(iet)
+
+
+def test_petsc_functions():
+    """
+    Test C++ support for PETScFunctions.
+    """
+    grid = Grid((2, 2))
+    x, y = grid.dimensions
+
+    ptr0 = PETScFunction(name='ptr0', dtype=np.float32, dimensions=grid.dimensions,
```

Comment:
Would it be worth checking all the datatypes here, so `np.float64`, `np.int64` etc?